find_posts.py

```diff
@@ -890,8 +1089,160 @@ def __len__(self):
+        if new_server is not None:
+            if new_server == server:
+                log(f'host-meta for {server} did not get a new server.')
+                return None
+            else:
+                return get_nodeinfo(new_server, seen_hosts, True)
+        else:
+            return None
+
+    if resp.status_code == 200:
+        try:
+            nodeInfo = resp.json()
+            for link in nodeInfo['links']:
+                if link['rel'] in [
+                    'http://nodeinfo.diaspora.software/ns/schema/2.0',
+                    'http://nodeinfo.diaspora.software/ns/schema/2.1',
+                ]:
+                    nodeLoc = link['href']
+                    break
+        except Exception as ex:
+            log(f'error getting server {server} info from well-known node info. Exception: {ex}')
+            return None
+    else:
+        log(f'Error getting well-known host node info for {server}. Status Code: {resp.status_code}')
+        return None
+
+    if nodeLoc is None:
+        log(f'could not find link to node info in well-known nodeinfo of {server}')
+        return None
+
+    # regrab server from nodeLoc, again in the case of different display and web domains
+    match = re.match(
+        r"https://(?P<server>[^/]+)/", nodeLoc
+    )
+    if match is None:
+        log(f"Error getting web server name from {server}.")
+        return None
+
+    server = match.group('server')
+
+    # return early if the web domain has been seen previously (in cases with host-meta lookups)
+    if server in seen_hosts:
+        return seen_hosts[server]
+
+    try:
+        resp = get(nodeLoc, timeout = 30)
+    except Exception as ex:
+        log(f"Error getting host node info for {server}. Exception: {ex}")
+        return None
+
+    if resp.status_code == 200:
+        try:
+            nodeInfo = resp.json()
+            if 'activitypub' not in nodeInfo['protocols']:
+                log(f'server {server} does not support activitypub, skipping')
+                return None
+            return {
+                'webserver': server,
+                'software': nodeInfo['software']['name'],
+                'version': nodeInfo['software']['version'],
+                'rawnodeinfo': nodeInfo,
+            }
+        except Exception as ex:
+            log(f'error getting server {server} info from nodeinfo. Exception: {ex}')
+            return None
+    else:
+        log(f'Error getting host node info for {server}. Status Code: {resp.status_code}')
+        return None
+
+def get_server_info(server, seen_hosts):
+    if server in seen_hosts:
+        serverInfo = seen_hosts.get(server)
+        if('info' in serverInfo and serverInfo['info'] == None):
+            return None
+        return serverInfo
+
+    nodeinfo = get_nodeinfo(server, seen_hosts)
+    if nodeinfo is None:
+        seen_hosts.add(server, {
+            'info': None,
+            'last_checked': datetime.now()
+        })
```

Comment:
Like I mentioned in earlier comments, some servers (like firefish.social) I've found sometimes either timeout or return 5XX errors for the nodeinfo request. We probably shouldn't silently skip that whole host for a month when that happens. Maybe either a super-short timeout period (like an hour or so) or just always try again for those hosts.

Reply:
I did actually address that in this commit - though I just realised that the comment is wrong there  😉

https://github.com/nanos/FediFetcher/pull/66/commits/3294a44f769a6b1aa067e6c7a1e5098ccf45341a